pkg/resources/podAntiAffinity.go

```diff
@@ -0,0 +1,143 @@
+)
+
+const (
+	// ZoneLabel is the label that specifies the zone where a node is
+	ZoneLabel = "topology.kubernetes.io/zone"
+	// AntiAffinityRequiredEnvVar is an environment variable that, when set to
+	// true, makes the product pod replicas use "required" anti affinity rules
+	// by AZ
+	AntiAffinityRequiredEnvVar = "FORCE_ZONE_DISTRIBUTION"
+)
+
+// MutateMultiAZAntiAffinity returns a PodTemplateMutation that sets the anti
+// affinity by AZ on the label labelMatch. It checks if the affinity rule is
+// required or not, and sets the required or preferred affinity based on it
+func MutateMultiAZAntiAffinity(ctx context.Context, client k8sclient.Client, labelMatch string) PodTemplateMutation {
+	isRequired, err := IsAntiAffinityRequired(ctx, client)
+	if err != nil {
+		isRequired = false
+	}
+
+	return func(obj metav1.Object, podTemplate *corev1.PodTemplateSpec) error {
+		labels := obj.GetLabels()
+		labelValue, ok := labels[labelMatch]
+		if !ok {
+			return fmt.Errorf("label %s not found in object", labelMatch)
+		}
+
+		podTemplate.Spec.Affinity = SelectAntiAffinityForCluster(isRequired, map[string]string{
+			labelMatch: labelValue,
+		})
+
+		return nil
+	}
+}
+
+// MultiAZAntiAffinityPreferred returns the affinity configuration to set the
+// preferred anti affinity by AZ on the given matchLabels
+func MultiAZAntiAffinityPreferred(matchLabels map[string]string) *corev1.Affinity {
+	return &corev1.Affinity{
+		PodAntiAffinity: &corev1.PodAntiAffinity{
+
+			PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{
+				{
+					PodAffinityTerm: corev1.PodAffinityTerm{
+						LabelSelector: &v1.LabelSelector{
+							MatchLabels: matchLabels,
+						},
+						TopologyKey: ZoneLabel,
+					},
+					Weight: 100,
+				},
+			},
+		},
+	}
+}
+
+// MultiAZAntiAffinityRequired returns the affinity configuration to set the
+// required anti affinity by AZ on the given matchLabels
+func MultiAZAntiAffinityRequired(matchLabels map[string]string) *corev1.Affinity {
+	return &corev1.Affinity{
+		PodAntiAffinity: &corev1.PodAntiAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+				{
+					LabelSelector: &v1.LabelSelector{
+						MatchLabels: matchLabels,
+					},
+					TopologyKey: ZoneLabel,
+				},
+			},
+		},
+	}
+}
+
+// SelectAntiAffinityForCluster returns the affinity configuration for a cluster
+// given whether the rule is required or preferred
+func SelectAntiAffinityForCluster(required bool, matchLabels map[string]string) *corev1.Affinity {
+	if required {
+		return MultiAZAntiAffinityRequired(matchLabels)
+	}
+
+	return MultiAZAntiAffinityPreferred(matchLabels)
+}
+
+// IsAntiAffinityRequired checks whether the anti affinity rule must be set
+// to required or preferred.
+//
+// It currently checks the value of the FORCE_ZONE_DISTRIBUTION bool env var
+func IsAntiAffinityRequired(_ context.Context, _ k8sclient.Client) (bool, error) {
```

Comment:
Are the function parameters needed here? I don't see them being used.

Reply:
Not at the moment. For now the result is based on an environment variable, but we might externalize it to a ConfigMap at some point, so I left the parameters that will be needed in the function signature